[TECH] Empêcher le reset du timer lors du refresh d'une page

celineung opened the pull request:
## :unicorn: Problème
Sur une épreuve timée, le temps est limitée.
Mais lorsque l'utilisateur rafraichit la page ou la quitte, le timer se reset et l'utilisateur dispose à nouveau de tout le temps nécessaire pour répondre à l'épreuve.

## :robot: Solution
1. Ajouter un listener sur l'évènement `beforeunload`: cet évènement est trigger lorsque la demande de rechargement de la page est faite. En utilisant `event.preventDefault`, on demande confirmation à l'utilisateur pour quitter la page.

2. Si l'utilisateur décide tout de même de quitter/rafraichir la page, alors l'event listener `unload` est triggered et on procède comme si l'utilisateur avait cliqué sur "Je passe".

## :rainbow: Remarques
- Je ne pense pas qu'il soit possible de customiser le message dans la popup de confirmation de quitter la page. J'ai l'impression que les navigateurs ont bloqué cette feature.

Autres pistes explorées:
- dans `router.js`, on écoutait déjà sur `routeDidChange`. Cet évènement peut prendre en paramètre la **transition** (permet savoir de quelle url on vient et vers quel url on va). Mais ça ne marche que si on navigue à l'intérieur de l'app. Si on quitte l'app, l'évènement n'est pas déclenché
- stocker un cookie contenant les épreuves timers en cours, mais les cookies peuvent être supprimés du navigateur

## :100: Pour tester

- Pour tester seulement une épreuve timée: rec1cSZ2hePondDBl.
- Pour tester dans le contexte d'une évaluation de compétence: la compétence "Gérer des données" a une épreuve timée dans ses premières questions (penser à reset la compétence).

Commit:
Fix not reset timer when page is refresh

Commit:
Add tests that verify user cannot reset timer on timed challenge

pix-service commented:
I'm deploying this PR to these urls:

- App: https://app-pr2591.review.pix.fr
- Orga: https://orga-pr2591.review.pix.fr
- Certif: https://certif-pr2591.review.pix.fr
- Admin: https://admin-pr2591.review.pix.fr
- API: https://api-pr2591.review.pix.fr/api/

 Please check it out!

alexandrecoin commented on mon-pix/app/components/challenge-item-generic.js:
```diff
@@ -15,6 +15,9 @@ export default class ChallengeItemGeneric extends Component {
   _timer = null;
+  _pageBeforeUnloadListener = null;
+  _pageUnloadListener = null;
+  _hasAbandOrValidateAnswer = false;
```
```suggestion
  _hasAbandonedOrValidatedAnswer = false;
```
**Suggestion: (non-blocking)** Renommage.

alexandrecoin commented on mon-pix/app/components/challenge-item-generic.js:
```diff
@@ -123,5 +133,29 @@ export default class ChallengeItemGeneric extends Component {
+    this._addQuitTimedChallengeListeners();
+  }
+
+  _removeEventListeners() {
+    if (this._pageBeforeUnloadListener) {
+      window.removeEventListener('beforeunload', this._pageBeforeUnloadListener);
+    }
+    if (this._pageUnloadListener) {
+      window.removeEventListener('unload', this._pageUnloadListener);
+    }
+  }
+
+  _addQuitTimedChallengeListeners() {
+    this._pageBeforeUnloadListener = (event) => {
+      event.preventDefault();
+    };
+    this._pageUnloadListener = (event) => {
+      event.preventDefault();
+      if (!this._hasAbandOrValidateAnswer) {
+        this.skipChallenge();
+      }
+    };
+    window.addEventListener('beforeunload', this._pageBeforeUnloadListener);
+    window.addEventListener('unload', this._pageUnloadListener);
```
```suggestion
  _addQuitTimedChallengeListeners() {
    this._pageBeforeUnloadListener = (event) => {
      event.preventDefault();
    };
    this._pageUnloadListener = (event) => {
      event.preventDefault();
      if (!this._hasAbandOrValidateAnswer) {
        this.skipChallenge();
      }
    };
    window.addEventListener('beforeunload', this._pageBeforeUnloadListener);
    window.addEventListener('unload', this._pageUnloadListener);
  }

  _removeEventListeners() {
    if (this._pageBeforeUnloadListener) {
      window.removeEventListener('beforeunload', this._pageBeforeUnloadListener);
    }
    if (this._pageUnloadListener) {
      window.removeEventListener('unload', this._pageUnloadListener);
    }
  }
```

**Suggestion (non-blocking) :** Inverser l'ordre des méthodes afin de mieux retranscrire l'ordre des évènements.

alexandrecoin reviewed (commented):
Fonctionnellement : A première vue, je peux encore avoir l'entièreté du temps à disposition si je rafraîchis la page. (cas de test de l'épreuve timée : rec1cSZ2hePondDBl )

Edit : Suite à la réponse de Mélanie, je confirme que j'ai testé sous Chrome également.

sbedeau commented on mon-pix/app/components/challenge-item-generic.js:
```diff
@@ -15,6 +15,9 @@ export default class ChallengeItemGeneric extends Component {
   _timer = null;
+  _pageBeforeUnloadListener = null;
+  _pageUnloadListener = null;
+  __hasAbandonedOrValidatedAnswer = false;
```
**question:** Pourquoi utiliser deux underscores ? 
```suggestion
  _hasAbandonedOrValidatedAnswer = false;
```

sbedeau commented on mon-pix/app/components/challenge-item-generic.js:
```diff
@@ -23,6 +26,11 @@ export default class ChallengeItemGeneric extends Component {
     }
   }
 
+  willDestroy() {
+    super.willDestroy(...arguments);
+    this._removeEventListeners();
+  }
```
**question:** Est-ce qu'on est sûr que la ligne `this._removeEventListeners()` sera bien appelée même après l'appel du super ? 
```suggestion
  willDestroy() {
    this._removeEventListeners();
    super.willDestroy(...arguments);
  }
```

sbedeau commented on mon-pix/app/components/challenge-item-generic.js:
```diff
@@ -113,6 +122,7 @@ export default class ChallengeItemGeneric extends Component {
       this.errorMessage = null;
       this.hasUserConfirmedWarning = false;
       this.isSkipButtonEnabled = false;
+      this.__hasAbandonedOrValidatedAnswer = true;
```
**suggestion:** On se trouve dans la méthode `skipChallenge`. J'aurais donc tendance à nommer comme suit :

```suggestion
      this.__hadSkippedOrValidatedAnswer = true;
```

sbedeau commented on mon-pix/app/components/challenge-item-generic.js:
```diff
@@ -123,5 +133,29 @@ export default class ChallengeItemGeneric extends Component {
+    };
+    this._pageUnloadListener = (event) => {
+      event.preventDefault();
+      if (!this.__hasAbandonedOrValidatedAnswer) {
+        this.skipChallenge();
+      }
```
**thought:** Si je comprend bien, ce listener sera toujours appelé lors d'une question timée et on a besoin du booléen pour décider si on skip ou non. Je trouve qu'il y a une forme de couplage et de complexité qui naît de ça. Est-ce qu'on ne saurait pas détecter autrement / plus explicitement qu'il y a une réponse enregistrée (validée ou passée) ? 

Peut-être simplement en renommant `__hasAbandonedOrValidatedAnswer` par `_hasAnswerBeenRecorded` ?

MelanieMEB reviewed (changes requested):
Cette approche ne semble marcher que sur Firefox (sous Chrome, aucun changement par rapport à avant).

Il va falloir aussi voir avec @PhilippineLoison pour le fonctionnel assez fort de cette PR, mais ça donne enfin de l'espoir pour réussir à gérer ce soucis :)

sbedeau commented on mon-pix/app/components/challenge-item-generic.js:
```diff
@@ -123,5 +133,29 @@ export default class ChallengeItemGeneric extends Component {
+  }
+
+  _addQuitTimedChallengeListeners() {
+    this._pageBeforeUnloadListener = (event) => {
+      event.preventDefault();
+    };
+    this._pageUnloadListener = (event) => {
+      event.preventDefault();
+      if (!this.__hasAbandonedOrValidatedAnswer) {
+        this.skipChallenge();
+      }
+    };
+    window.addEventListener('beforeunload', this._pageBeforeUnloadListener);
+    window.addEventListener('unload', this._pageUnloadListener);
```
**suggestion (non-blocking):** Les explications faites dans la description de la PR devraient pouvoir ressortir dans le code. Je pense qu'on y gagnerait à encapsuler le listener dans des méthodes qui explicitent l'intention pour les néophytes sur le sujet

```suggestion
_addQuitPromptListener() { 
    this._pageBeforeUnloadListener = (event) => {
      event.preventDefault();
    };
    window.addEventListener('beforeunload', this._pageBeforeUnloadListener);
}

_addOnConfirmedQuitPromptListener() {
	this._pageUnloadListener = (event) => {
      event.preventDefault();
      if (!this.__hasAbandonedOrValidatedAnswer) {
        this.skipChallenge();
      }
     } 
     window.addEventListener('unload', this._pageUnloadListener);
 } 
```

Les suggestion plus fines de nommage sont évidemment les bienvenues.

celineung replied:
Pour le premier `_addQuitPromptListener`: cela signifie que l'on met un listener sur la popup (à cause du mot `prompt`). Ce qui n'est pas le cas: on met un listener sur l'évènement rafraichir/quitter la page invoqué par l'utilisateur. Du coup, je suggèrerais juste `_addQuitListener`.

Commit:
Fix wait for post answers request

jonasgrilleres commented on high-level-tests/e2e/cypress/integration/pix-app/challenge.feature:
```diff
@@ -0,0 +1,12 @@
+    Lorsque je clique sur "Commencer"
+    Alors je vois l'épreuve "Combien font 2 * 5 ?"
+    Lorsque je rafraichis ma page
+    Alors je vois que l'épreuve a déjà été répondue
```
```suggestion
    Alors je vois qu'il a déjà été répondu à l'épreuve
```

jonasgrilleres commented on high-level-tests/e2e/cypress/support/step_definitions/assessment.js:
```diff
@@ -41,3 +41,7 @@ then(`j'ai bien répondu à {string}`, (challenge) => {
     .should('have.class', 'fa-check-circle').and('have.class', 'result-item__icon--green');
 });
+
+then(`je vois que l'épreuve a déjà été répondue`, () => {
```
```suggestion
then(`je vois qu'il a déjà été répondu à l'épreuve`, () => {
```

Commit:
Fix Chrome beforeunload call

By adding a event.returnValue, some browsers may display it as the popup message.

Commit:
[skip ci] WIP

celineung commented:
PR fermée car au moment de recharger la page, le POST vers /answers (pour signaler à l'api que l'épreuve a été abandonnée) n'est pas aboutie. Il aurait fallu dans l'event listener `pagehide`/`unload` faire une requete `sendBeacon` ou `XMLHttpRequest`, mais dans ce cas il faudra reformatter la donnée au format JSONApi + mettre à jour la donnée dans EmberData car on sortira du cadre Ember.
Une autre solution pour empêcher de reset les timers sera étudiée dans une autre PR.
